### PR TITLE
Configurable destinations for nbextensions.

### PR DESCRIPTION
**Pull request by jasongrout**
This allows for a directory in a package to be called 'nbextensions', but be installed to an nbextension named after the package.  This addresses #6939.

### CODE REVIEW

**Comment by takluyver** [comment]
LGTM, but I'd like to see a test in `IPython.html.tests.test_nbextensions`.

**Comment by jasongrout** [comment]
There are a few bugs that I'm working through.  I'll fix those and put a few tests in.

**Comment by jasongrout** [comment]
Did you know that install_nbextension actually copies the files over every single time on python 2.7.9 on a modern linux?  I think it's related to this bug: http://bugs.python.org/issue12904.  Basically, getting the times gets them in nanosecond precision, but setting them only sets them in microsecond precision.  Here is an example, `s` is a filename.
```python
In [43]: oldtimes=(os.stat(s).st_atime, os.stat(s).st_mtime)

In [44]: os.utime(s, (oldtimes))

In [45]: newtimes=(os.stat(s).st_atime, os.stat(s).st_mtime)

In [46]: oldtimes==newtimes  
Out[46]: False

In [47]: oldtimes[0]-newtimes[0]
Out[47]: 9.5367431640625e-07

In [48]: oldtimes[1]-newtimes[1]
Out[48]: 9.5367431640625e-07
```

**Comment by jasongrout** [comment]
I'm going to add a fudge factor to checking the mtime to get around this issue.

**Comment by jasongrout** [comment]
@takluyver - I got the last doctest working because I changed the names to not use weird unicode characters.  I'm not sure what is going wrong on python 2 with the unicode characters (see https://travis-ci.org/ipython/ipython/jobs/47975940#L340).  Do you happen to see what is going on and how to fix the test (or the code) in the case where we have a weird unicode filename?

**Comment by takluyver** [comment]
I suspect that you just need the filename to be a unicode literal. i.e. `u'ƒile.js'` instead of `'ƒile.js'`.

**Comment by jasongrout** [comment]
Ah, right, of course.  We'll see if Travis is happy now...

**Comment by jasongrout** [comment]
Looks like things are good.  @takluyver, I think this is ready for review.

I noticed that there is a bit of code complexity because you can specify a list of extensions.  My guess is that the vast majority of the time, you're only installing one extension.  In the rare exception that you'd be installing multiple extensions, I don't think it's that onerous to have people call the function multiple times.  Do you mind if I simplify the function to just handle installing one extension?

**Comment by jasongrout** [comment]
(simplify to handle just one extension on another PR.  I think this PR is feature-frozen at this point)

**Comment by takluyver** [comment]
Thanks, will review later this afternoon.

I'd say you're free to make a PR simplifying this function to only install one extension at a time, then we'll discuss whether we prefer the extra functionality or the technical simplicity.

File: IPython/html/nbextensions.py [hunk 1]
```diff
@@ -74,8 +77,10 @@ def _maybe_copy(src, dest, verbose=1):
     if _should_copy(src, dest, verbose):
         if verbose >= 1:
             print("copying %s -> %s" % (src, dest))
-        shutil.copy2(src, dest)
-
+        try:
+            shutil.copy2(src, dest)
+        except IOError as e:
+            print(str(e), file=sys.stderr)
```
Comment (takluyver) [inline IPython/html/nbextensions.py]: Why is this error being printed but suppressed?

**Comment by takluyver** [comment]
Other than that, :+1:

File: IPython/html/nbextensions.py (hunk 1 above)
Comment (jasongrout) [inline IPython/html/nbextensions.py]: It is printed for informational purposes, but suppressed to try to get as much of the extension as possible copied over.  Would you rather that we didn't suppress the error?

**Comment by jasongrout** [comment]
As for the install-single-extension PR, why don't we decide which we'd rather have before I spend the time doing it.

File: IPython/html/nbextensions.py (hunk 1 above)
Comment (takluyver) [inline IPython/html/nbextensions.py]: Yes. Exceptions tell the calling code that something is wrong, which may be able to do something about it. Additionally, if copying one file has failed:

1. It's very likely that copying the next one will fail, so for a large extension, the user will see a long line of intimidating errors.
2. It's very likely that the extension won't work correctly, so there's no use pretending to the caller that everything's fine.

**Comment by takluyver** [comment]
Don't feel you have to write it, but if we're weighing up the relative merits of simpler code versus more features, and we already have the more featureful version implemented, then it's easier to make the case for simplification if it's clear how much simpler the code would be.

File: IPython/html/nbextensions.py (hunk 1 above)
Comment (jasongrout) [inline IPython/html/nbextensions.py]: Fair enough.  Those weren't the case for my example extension, but I agree that those are the more likely cases.  I'll take out the error trapping.

**Comment by jasongrout** [comment]
@takluyver: I fixed the trapping error issue - can you take a look at this again?

**Comment by takluyver** [comment]
Looks OK to me; just re-running a Travis test group that failed, but I'm pretty sure it's unrelated.

**Pull request merged**